File: app-ios-nativa/Dominio/Pruebas/PruebasDominio/PruebasDuplicados.swift
import Foundation
import Testing

@testable import Dominio

@Suite("Normalizar una URL para compararla")
struct PruebasNormalizarUrl {
    @Test("http y https son la misma página")
    func esquemaDaIgual() {
        #expect(Duplicados.esLaMisma("http://ejemplo.com/a", "https://ejemplo.com/a"))
    }

    @Test("el «www», la barra final y el ancla no cuentan")
    func adornosNoCuentan() {
        #expect(Duplicados.esLaMisma("https://www.ejemplo.com/a/", "https://ejemplo.com/a#seccion"))
    }

    @Test("los parámetros de seguimiento se descartan")
    func seguimientoFuera() {
        #expect(
            Duplicados.esLaMisma(
                "https://ejemplo.com/a?utm_source=boletin&fbclid=xyz",
                "https://ejemplo.com/a"
            )
        )
    }

    @Test("pero los parámetros que identifican el contenido sí cuentan")
    func parametrosDeContenido() {
        #expect(
            !Duplicados.esLaMisma("https://ejemplo.com/ver?id=1", "https://ejemplo.com/ver?id=2"))
    }

    @Test("el orden de los parámetros da igual")
    func ordenDaIgual() {
        #expect(
            Duplicados.esLaMisma("https://ejemplo.com/?b=2&a=1", "https://ejemplo.com/?a=1&b=2"))
    }

    @Test("la ruta distingue mayúsculas: hay servidores donde no es lo mismo")
    func rutaDistingueMayusculas() {
        #expect(!Duplicados.esLaMisma("https://ejemplo.com/Uno", "https://ejemplo.com/uno"))
    }

    @Test("lo que no es una URL se compara tal cual, sin inventar duplicados")
    func loQueNoEsUrl() {
        #expect(Duplicados.normalizar("  esto no es una url  ") == "esto no es una url")
        #expect(!Duplicados.esLaMisma("esto no es una url", "ni esto tampoco"))
    }

    @Test("dos cosas que no son URL tampoco se confunden entre sí por estar vacías")
    func sinAnfitrion() {
        // El fallo que esto caza es el de la app de Expo: allí el `URL` de
        // React Native devolvía anfitrión vacío para cualquier cosa, así que
        // todo lo que no fuera una URL normalizaba igual y se tomaba por
        // duplicado de lo anterior.
        #expect(Duplicados.normalizar("hola") != Duplicados.normalizar("adios"))
    }
}

@Suite("Buscar un duplicado en lo ya guardado")
struct PruebasBuscarDuplicado {
    private let guardados = [
        Elemento(id: "e1", url: "https://www.xataka.com/basics/alternativas-pocket"),
        Elemento(id: "e2", url: "https://ejemplo.com/otro"),
    ]

    @Test("encuentra el que ya estaba, aunque venga escrito de otra forma")
    func loEncuentra() {
        let encontrado = Duplicados.buscar(
            en: guardados,
            url: "http://xataka.com/basics/alternativas-pocket/?utm_source=twitter"
        )

        #expect(encontrado?.id == "e1")
    }

    @Test("una URL nueva no es duplicado de nada")
    func urlNueva() {
        #expect(Duplicados.buscar(en: guardados, url: "https://ejemplo.com/nuevo") == nil)
    }

    @Test("lo borrado no cuenta: volver a guardarlo es un alta normal")
    func loBorradoNoCuenta() {
        let conBorrado = [Elemento(id: "e3", url: "https://ejemplo.com/ido", borrado: true)]

        #expect(Duplicados.buscar(en: conBorrado, url: "https://ejemplo.com/ido") == nil)
    }
}

File: app-android/dominio/src/test/kotlin/com/jmortizsilva/guardarenlaces/dominio/PruebasDuplicados.kt
package com.jmortizsilva.guardarenlaces.dominio

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue

class PruebasNormalizarUrl {
    @Test
    fun `http y https son la misma pagina`() {
        assertTrue(Duplicados.esLaMisma("http://ejemplo.com/a", "https://ejemplo.com/a"))
    }

    @Test
    fun `el www, la barra final y el ancla no cuentan`() {
        assertTrue(
            Duplicados.esLaMisma("https://www.ejemplo.com/a/", "https://ejemplo.com/a#seccion")
        )
    }

    @Test
    fun `los parametros de seguimiento se descartan`() {
        assertTrue(
            Duplicados.esLaMisma(
                "https://ejemplo.com/a?utm_source=boletin&fbclid=xyz",
                "https://ejemplo.com/a",
            )
        )
    }

    @Test
    fun `pero los parametros que identifican el contenido si cuentan`() {
        assertFalse(
            Duplicados.esLaMisma("https://ejemplo.com/ver?id=1", "https://ejemplo.com/ver?id=2")
        )
    }

    @Test
    fun `el orden de los parametros da igual`() {
        assertTrue(
            Duplicados.esLaMisma("https://ejemplo.com/?b=2&a=1", "https://ejemplo.com/?a=1&b=2")
        )
    }

    @Test
    fun `dos parametros con la misma clave conservan su orden`() {
        assertFalse(
            Duplicados.esLaMisma("https://ejemplo.com/?a=1&a=2", "https://ejemplo.com/?a=2&a=1")
        )
    }

    @Test
    fun `la ruta distingue mayusculas, hay servidores donde no es lo mismo`() {
        assertFalse(Duplicados.esLaMisma("https://ejemplo.com/Uno", "https://ejemplo.com/uno"))
    }

    @Test
    fun `un mas en la consulta es un mas, no un espacio`() {
        // `URLDecoder` de Java es para formularios y lo convertiría en espacio: «c++» y «c  »
        // pasarían a ser la misma búsqueda.
        assertFalse(
            Duplicados.esLaMisma("https://ejemplo.com/?q=c++", "https://ejemplo.com/?q=c%20%20")
        )
    }

    @Test
    fun `lo que no es una URL se compara tal cual, sin inventar duplicados`() {
        assertEquals("esto no es una url", Duplicados.normalizar("  esto no es una url  "))
        assertFalse(Duplicados.esLaMisma("esto no es una url", "ni esto tampoco"))
    }

    @Test
    fun `dos cosas que no son URL tampoco se confunden entre si por estar vacias`() {
        // El fallo que esto caza es el de la app de Expo: el `URL` de React Native devolvía
        // anfitrión vacío para cualquier cosa, y todo lo que no fuera una URL se tomaba por
        // duplicado de lo anterior.
        assertNotEquals(Duplicados.normalizar("hola"), Duplicados.normalizar("adios"))
    }
}

class PruebasBuscarDuplicado {
    private val guardados =
        listOf(
            Elemento(id = "e1", url = "https://www.xataka.com/basics/alternativas-pocket"),
            Elemento(id = "e2", url = "https://ejemplo.com/otro"),
        )

    @Test
    fun `encuentra el que ya estaba, aunque venga escrito de otra forma`() {
        val encontrado =
            Duplicados.buscar(
                guardados,
                "http://xataka.com/basics/alternativas-pocket/?utm_source=twitter",
            )

        assertEquals("e1", encontrado?.id)
    }

    @Test
    fun `una URL nueva no es duplicado de nada`() {
        assertNull(Duplicados.buscar(guardados, "https://ejemplo.com/nuevo"))
    }

    @Test
    fun `lo borrado no cuenta, volver a guardarlo es un alta normal`() {
        val conBorrado =
            listOf(Elemento(id = "e3", url = "https://ejemplo.com/ido", borrado = true))

        assertNull(Duplicados.buscar(conBorrado, "https://ejemplo.com/ido"))
    }
}
